File: Stack/FlattenBinaryTreetoLinkedList_114.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
 // Pre-order traversal
class Solution {
    public void flatten(TreeNode root) {
        Stack<TreeNode> stack = new Stack<>();
        TreeNode p = root;
        while (p != null || !stack.empty()){
            if (p.right != null){
                    stack.push(p.right);                 
                }
            if (p.left != null){ 
                p.right = p.left;
                p.left = null; 
            }
            else if (!stack.empty()){
                p.right = stack.pop();
            }
            p = p.right;            
        }
    }
}

File: BinaryTree/FlattenBinaryTreetoLinkedList_114.py
# beat 100% Py3
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        # preorder traversal
        if root == None: return
        stack = [root]
        while stack:
            cur = stack.pop()
            if cur.right:
                stack.append(cur.right)
            if cur.left:
                stack.append(cur.left)
            cur.left = None
            cur.right = stack[-1] if len(stack) > 0 else None
